### PR TITLE
Stable: Improve math roundings

### PR DESCRIPTION
**Pull request by facuspagnuolo**

### CODE REVIEW

File: pkg/core/contracts/pools/BasePool.sol [hunk 1]
```diff
@@ -453,7 +453,7 @@ abstract contract BasePool is IBasePool, BasePoolAuthorization, BalancerPoolToke
      */
     function _addSwapFeeAmount(uint256 amount) internal view returns (uint256) {
         // This returns amount + fee amount, so we round up (favoring a higher fee amount).
-        return amount.divUp(_swapFeePercentage.complement());
+        return amount.divUp(FixedPoint.ONE.sub(_swapFeePercentage));
```
Comment (nventuro) [inline pkg/core/contracts/pools/BasePool.sol:456]: This is just more clear, but it doesn't change the behavior, right?

File: pkg/core/contracts/pools/stable/StableMath.sol [hunk 2]
```diff
@@ -106,7 +111,7 @@ contract StableMath {
 
         // Amount out, so we round down overall.
 
-        uint256 invariant = _calculateInvariant(amplificationParameter, balances);
+        uint256 invariant = _calculateInvariant(amplificationParameter, balances, true);
```
Comment (nventuro) [inline pkg/core/contracts/pools/stable/StableMath.sol:114]: The comment says we're rounding down, but we're rounding up.

File: pkg/core/contracts/pools/stable/StableMath.sol [hunk 3]
```diff
@@ -196,43 +198,42 @@ contract StableMath {
...
+        uint256 currentInvariant = _calculateInvariant(amp, balances, true);
+        uint256 newInvariant = _calculateInvariant(amp, newBalances, false);
+        uint256 invariantRatio = newInvariant.divDown(currentInvariant);
 
-        // return amountBPTOut
-        return bptTotalSupply.mulDown(newInvariant.divDown(currentInvariant).sub(FixedPoint.ONE));
+        if (invariantRatio > FixedPoint.ONE) {
+            return bptTotalSupply.mulDown(invariantRatio.sub(FixedPoint.ONE));
+        } else {
+            return 0;
```
Comment (nventuro) [inline pkg/core/contracts/pools/stable/StableMath.sol:235]: This should only happen due to rounding, correct? Can we have a small comment explaining that if this is the case?

File: pkg/core/contracts/pools/stable/StableMath.sol [hunk 4]
```diff
@@ -357,34 +355,40 @@ contract StableMath {
         uint256 bptTotalSupply,
         uint256 swapFeePercentage
     ) internal pure returns (uint256) {
+        // Token out, so we round down overall.
+
         // Get the current invariant
-        uint256 currentInvariant = _calculateInvariant(amp, balances);
+        uint256 currentInvariant = _calculateInvariant(amp, balances, true);
```
Comment (nventuro) [inline pkg/core/contracts/pools/stable/StableMath.sol:361]: If we're rounding down, why is the invariant rounded up? It'd be good to have a short comment about this

File: pkg/core/contracts/pools/stable/StablePool.sol [hunk 5]
```diff
@@ -409,23 +413,34 @@ contract StablePool is BaseGeneralPool, StableMath {
     }
 
     function _invariantAfterJoin(uint256[] memory balances, uint256[] memory amountsIn) private view returns (uint256) {
-        for (uint256 i = 0; i < _getTotalTokens(); ++i) {
-            balances[i] = balances[i].add(amountsIn[i]);
-        }
-
-        return StableMath._calculateInvariant(_amplificationParameter, balances);
+        _mutateAmounts(balances, amountsIn, FixedPoint.add);
+        // This invariant is used only to calculate the due protocol fees, then we round up
```
Comment (nventuro) [inline pkg/core/contracts/pools/stable/StablePool.sol:417]: Shouldn't protocol fees be rounded down?

File: pkg/core/contracts/pools/stable/StablePool.sol [hunk 6]
```diff
@@ -434,6 +449,10 @@ contract StablePool is BaseGeneralPool, StableMath {
      */
     function getRate() public view returns (uint256) {
         (, uint256[] memory balances, ) = getVault().getPoolTokens(getPoolId());
-        return StableMath._calculateInvariant(_amplificationParameter, balances).divDown(totalSupply());
+
+        // When calculating the current BPT rate, we may not have paid the protocol fees, therefore
+        // the invariant should be smaller than it's current value. Then, we round down overall.
```
Comment (nventuro) [inline pkg/core/contracts/pools/stable/StablePool.sol:454]:
```suggestion
        // the invariant should be smaller than its current value. Then, we round down overall.
```

**Review by nventuro — COMMENTED** [review]
Excelent :raised_hands: Thank you very much for updating the comments and bringing this more in line with the Weighted Pool!

I did an initial pass asking a few questions about things that seemed odd, I'll do another more detailed pass later today.

File: pkg/core/contracts/pools/BasePool.sol (hunk 1 above)
Comment (facuspagnuolo) [reply pkg/core/contracts/pools/BasePool.sol]: We did change the behavior, but the swap fee pct is guaranteed to be lower than or equal to 50%

File: pkg/pool-stable/contracts/StablePool.sol [hunk 7]
```diff
@@ -410,23 +414,34 @@ contract StablePool is BaseGeneralPool, StableMath {
     }
 
     function _invariantAfterJoin(uint256[] memory balances, uint256[] memory amountsIn) private view returns (uint256) {
-        for (uint256 i = 0; i < _getTotalTokens(); ++i) {
-            balances[i] = balances[i].add(amountsIn[i]);
-        }
-
-        return StableMath._calculateInvariant(_amplificationParameter, balances);
+        _mutateAmounts(balances, amountsIn, FixedPoint.add);
+        // This invariant is used only to compute the final balance when calculating the protocol fees, then we round up
```
Comment (nventuro) [inline pkg/pool-stable/contracts/StablePool.sol:418]:
```suggestion
        // This invariant is used only to compute the final balance when calculating the protocol fees. These are rounded
        // down, so we round the invariant up.
```

File: pkg/pool-stable/contracts/StablePool.sol [hunk 8]
```diff
@@ -410,23 +414,34 @@ contract StablePool is BaseGeneralPool, StableMath {
...
+        return StableMath._calculateInvariant(_amplificationParameter, balances, true);
     }
 
     function _invariantAfterExit(uint256[] memory balances, uint256[] memory amountsOut)
         private
         view
         returns (uint256)
     {
+        _mutateAmounts(balances, amountsOut, FixedPoint.sub);
+        // This invariant is used only to compute the final balance when calculating the protocol fees, then we round up
```
Comment (nventuro) [inline pkg/pool-stable/contracts/StablePool.sol:428]:
```suggestion
        // This invariant is used only to compute the final balance when calculating the protocol fees. These are rounded
        // down, so we round the invariant up.
```

**Review by nventuro — APPROVED** [review]

**Pull request merged**